File: Array/removeDuplicatesFromSortedArray.py
#Author : Yuan Wang
#Date   : 2018-05-26

#********************************************************************************** 
# 
# Given a sorted array, remove the duplicates in place such that each element appear 
# only once and return the new length.
# 
# Do not allocate extra space for another array, you must do this in place with constant memory.
# 
# For example,
# Given input array A = [1,1,2],
# 
# Your function should return length = 2, and A is now [1,2].
# 
#               
#**********************************************************************************/

#worked well with less time
def removeDuplicates(nums):
	"""
	:type nums: List[int]
	:rtype: int
	"""
	k=0  
	for i in range(1,len(nums)):  
		if nums[i] != nums[k]:  
			k+=1  
			nums[k] = nums[i]  
	  
	del nums[k+1:len(nums)]

#worked well but slow for large array	
def removeDuplicatesB(nums):
	for i in nums:
		repeat=nums.count(i)
		if repeat > 1:
			for j in range(repeat-1):
				nums.remove(i)

#worked well but slow for large array	
def removeDuplicatesB(nums):
	for i in nums:
		while(nums.count(i)>1):
			nums.remove(i)

A=[1,1,2,3,3,3,4,4,5]
removeDuplicates(A)
print(A)

File: Array/removeDuplicatesFromSortedArray.java
/*
#Author : Yuan Wang
#Date   : 2018-05-26

Algorithm

Since the array is already sorted, we can keep two pointers ii and jj, 
where ii is the slow-runner while jj is the fast-runner. As long as 
nums[i] = nums[j]nums[i]=nums[j], we increment jj to skip the duplicate.

When we encounter nums[j] \neq nums[i]nums[j]≠nums[i], the duplicate run
has ended so we must copy its value to nums[i + 1]nums[i+1]. ii is then 
incremented and we repeat the same process again until jj reaches the end of array.

Complexity analysis

Time complextiy : O(n)O(n). Assume that nn is the length of array.
Each of ii and jj traverses at most nn steps.

Space complexity : O(1)O(1).
*/


import java.util.Arrays;

public class removeDuplicatesFromSortedArray {
    public static void main(String[] args){
        int array[] = {1,1,2,3,3,3,4,4,5};
        int count=0;
        count=removeDuplicates(array);
        System.out.println(count);
        
    }

    public static int removeDuplicates(int[] nums){
        if (nums.length == 0) return 0;
        int i = 0;
        for (int j = 1; j < nums.length; j++) {
            if (nums[j] != nums[i]) {
                i++;
                nums[i] = nums[j];
        }
    }
    return i + 1;
    }
}